impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java

```diff
@@ -698,64 +831,86 @@ public String toString() {
     checkKey(key);
 
     final long now = timeSource.getTimeMillis();
-    OnHeapValueHolder<V> computeResult = map.compute(key, new BiFunction<K, OnHeapValueHolder<V>, OnHeapValueHolder<V>>() {
-      @Override
-      public OnHeapValueHolder<V> apply(K mappedKey, OnHeapValueHolder<V> mappedValue) {
-        if (mappedValue != null && mappedValue.isExpired(now, TimeUnit.MILLISECONDS)) {
-          onExpiration(mappedKey, mappedValue);
-          mappedValue = null;
-        }
-        
-        V existingValue = mappedValue == null ? null : mappedValue.value();
-        V computedValue = mappingFunction.apply(mappedKey, existingValue);        
-        if (computedValue == null) {
-          return null;
-        } else if ((eq(existingValue, computedValue)) && (! replaceEqual.apply())) {
+    try {
+      OnHeapValueHolder<V> computeResult = map.compute(key, new BiFunction<K, OnHeapValueHolder<V>, OnHeapValueHolder<V>>() {
+        @Override
+        public OnHeapValueHolder<V> apply(K mappedKey, OnHeapValueHolder<V> mappedValue) {
+          if (mappedValue != null && mappedValue.isExpired(now, TimeUnit.MILLISECONDS)) {
+            onExpiration(mappedKey, mappedValue);
+            mappedValue = null;
+          }
+
+          V existingValue = mappedValue == null ? null : mappedValue.value();
+          V computedValue = mappingFunction.apply(mappedKey, existingValue);
+          if (computedValue == null) {
+            return null;
+          } else if ((eq(existingValue, computedValue)) && (!replaceEqual.apply())) {
+            if (mappedValue != null) {
+              setAccessTimeAndExpiry(key, mappedValue, now);
+            }
+            return mappedValue;
+          }
+
+          checkValue(computedValue);
           if (mappedValue != null) {
-            setAccessTimeAndExpiry(key, mappedValue, now);
+            long expirationTime = mappedValue.expirationTime(OnHeapValueHolder.TIME_UNIT);
+            return newUpdateValueHolder(key, existingValue, computedValue, now, expirationTime);
+          } else {
+            return newCreateValueHolder(key, computedValue, now);
           }
-          return mappedValue;
-        }
-        
-        checkValue(computedValue);
-        if (mappedValue != null) {
-          long expirationTime = mappedValue.expirationTime(OnHeapValueHolder.TIME_UNIT);
-          return newUpdateValueHolder(key, existingValue, computedValue, now, expirationTime);
-        } else {
-          return newCreateValueHolder(key, computedValue, now);
         }
+      });
+      return enforceCapacityIfValueNotNull(computeResult);
+    } catch (CachePassThroughException cpte) {
+      Throwable cause = cpte.getCause();
```

Comment:
These few lines are repeated all over the place and in the `AbstractOffHeapStore` as well.
So they should be extracted. What is also interesting is that none of this requires the context of `this` class, so it would be a valid candidate for a method on `CachePassThroughException`.